Simplenote/SimplenoteAppDelegate.m

```diff
@@ -569,19 +570,45 @@ + (SimplenoteAppDelegate *)sharedDelegate
 #pragma mark - Actions
 
 - (IBAction)signOutAction:(id)sender
+{
+    // Safety first: Check for unsynced notes before they are deleted!
+    if ([StatusChecker hasUnsentChanges:self.simperium] == false)  {
+        [self signOut];
+        return;
+    }
+
+    NSAlert *alert = [[NSAlert alloc] init];
+    [alert addButtonWithTitle:NSLocalizedString(@"Sign Out", @"Sign out of the app")];
+    [alert addButtonWithTitle:NSLocalizedString(@"Visit Web App", @"Visit app.simplenote.com in the browser")];
+    [alert addButtonWithTitle:NSLocalizedString(@"Cancel", @"Cancel the action")];
+    [alert setMessageText:NSLocalizedString(@"Unsynced Notes Detected", @"Alert title displayed in when an account has unsynced notes")];
+    [alert setInformativeText:NSLocalizedString(@"Signing out will delete any unsynced notes. You can verify your synced notes by signing in to the Web App.", @"Alert message displayed when an account has unsynced notes")];
+    [alert setAlertStyle:NSAlertStyleCritical];
+
+    [alert beginSheetModalForWindow:[self window] completionHandler:^(NSInteger result) {
+        if (result == NSAlertSecondButtonReturn) {
+            NSURL *linkUrl = [NSURL URLWithString:@"https://app.simplenote.com"];
+            [[NSWorkspace sharedWorkspace] openURL:linkUrl];
+        } else if (result == NSAlertFirstButtonReturn) {
+            [self signOut];
+        }
```

Comment:
is this being passed into the `alert` as a body of code to run or is it returning a value as a parameter to `alert`?

Reply:
I'm not super clear on your question, but the result constants like `NSAlertFirstButtonReturn` represent which button was clicked in the alert.

Reply:
@dmsnell That block of code is an anonymous function (Closure in Swift, Block in ObjC). And precisely, as you mentioned, it's passed to the alert, to be executed whenever it's dismissed.

**ObjC Block:**
```
^(Type parameter) {
// ..
}
```

**Swift Closure:**
```
{ parameter in
// ..
}
```

Swift can infer the parameter's type, so, as long as it's defined in the receiver function (or property, depending on what you're doing!) you can get away with a compact syntax, like that one.

Reply:
thanks @jleandroperez - TIL about `^(…) { … }` :smile:

Reply:
Anytime sir!! happy to see you picking up iOS!!!